Netflix/pollyjs
# fix: use watch strategy

NB: Fixes #232

NB: submitting this for a gut-check / as a proof of concept. There's a lot of style / cleanup work that needs to be done before this is ready for a line-by-line.

## Motivation and Context

Let's say you're a new developer working on Polly. You check out the repo, install deps and run `$ yarn test`. This kicks off the following steps:

1. `yarn clean`
1. `yarn build` for server & utils
1. `yarn build` and `test:build` for ~12 sundry sub-packages
1. `yarn testem`, which kicks off the test runner.

This is a development pain point for the following reasons:

1. These tasks are run as a pre-hook inside `testem`, which makes the testing ui unresponsive for the ~2m it take to run the build.
1. Users are forced to re-run the build process whenever they modify either the tests or the application.
1. Hitting enter in the `testem` ui re-triggers the hooks, which can saturate your CPU if you're impatient.

## Description

While this required a lot of trial and error, the changes are actually fairly contained:

1. `testem` no longer runs any hooks. Running long tasks makes the UI unstable.
1. `testem` now only re-runs on changes to build artifacts (code and tests) instead of the source
1. Every sub-package now has a `yarn watch` command that triggers a watch on its code and tests
1. The root `package.json` now has a `yarn watch` command that runs all of the sub-packages' watch commands in parallel.

## Testing instructions

* in one tab, run `yarn watch` and wait for it to settle
* run `yarn test` in another tab
* trigger a failure by, for example, adding "`expect(false).to.equal(true)`" line 6 of `tests/integration/adapter-browser-tests.js`
* observe that polly selectively re-builds the select file and re-triggers the tests

## Tradeoffs, deficiencies and future work

### General concerns
1. I decided to go with this approach because it was the only one I could find that didn't require me to take a chainsaw to the actual configs.
1. The best approach I can find would be to use (https://github.com/rollup/rollup/pull/1389)(https://github.com/rollup/rollup/pull/1389) approach. tl;dr you can pass an array of rollup configs to a single rollup instance. This would allow us to have a single rollup instance watch every sub-package instead of having 15 separate processes, but trying to aggregate the final configs turned out to be really painful (because of liberal use of `process.cwd` among other things).

### Developer workflow
As implemented, this requires developers to manually kick off watch, see when it settles and then start testem. I personally prefer this approach, but I suspect something more streamlined will be required. So, I'd propose the following:

1. `yarn test` should run `yarn clean && yarn build && yarn testem` and then launch `yarn watch` in the background. I'm not exactly sure how this would be implemented, but quitting testem would *have* to terminate the watchers.
1. Developers who want a little more control can use `yarn test:watch` and `yarn build:watch` to use the workflow I'm submitting with -- i.e. complete separation of test running and build, which imo is preferable if for example you are booting testem and know that an incremental build is sufficient. (As an aside, an incremental build is actually *always* sufficient since it can build from scratch -- the problem is that there's no obvious way to block testem until the incremental build has finished, aside from writing something to block that process until every artifact is present).

## Merge blockers

- [ ] style check for commit, code changes and so forth
- [ ] `test/integration/adapter-node-tests.js` seems to not be properly run on master?
- [ ] look into the ember submodule -- seems to be excluded
- [ ] talk to polly team and agree on naming script commands
- [ ] remove redundant script commands

## Types of Changes


## Checklist

- [x] My commits and the title of this PR follow the [Conventional Commits Specification](https://www.conventionalcommits.org).
- [x] I have read the [contributing guidelines](https://github.com/Netflix/pollyjs/blob/master/CONTRIBUTING.md).
